File: recursion/javascript/solutions/chunksOf.js
/**
 * Solution for chunksOf.
 */

const { EMPTY_LIST, Node, isEmpty, prepend, unprepend, toArray } = require('./linkedList');
const { take } = require('./take');
const { drop } = require('./drop');

function chunksOf(n, list) {
  if (isEmpty(list)) {
    return EMPTY_LIST;
  }

  let chunk = take(n, list);
  let remaining = drop(n, list);

  return prepend(chunk, chunksOf(n, remaining));
}

if (require.main === module) {
  let result = chunksOf(2, Node.fromArray([1, 2, 3, 4, 5]));
  let chunks = toArray(result).map(toArray);
  console.assert(JSON.stringify(chunks) === '[[1,2],[3,4],[5]]');

  let result2 = chunksOf(3, Node.fromArray([1, 2]));
  let chunks2 = toArray(result2).map(toArray);
  console.assert(JSON.stringify(chunks2) === '[[1,2]]');

  console.assert(JSON.stringify(toArray(chunksOf(2, EMPTY_LIST))) === '[]');

  console.log('All tests passed!');
}

module.exports = { chunksOf };

File: recursion/python/solutions/chunks_of.py
"""Solution for chunks_of."""

from linked_list import EMPTY_LIST, Node, is_empty, prepend, unprepend, to_python_list
from take import take
from drop import drop


def chunks_of(n, lst):
    """
    Group elements into sublists of size n.

    chunks_of(2, 1 -> 2 -> 3 -> 4 -> 5) => (1 -> 2) -> (3 -> 4) -> (5)
    chunks_of(3, 1 -> 2) => (1 -> 2)
    """
    if is_empty(lst):
        return EMPTY_LIST

    chunk = take(n, lst)
    remaining = drop(n, lst)

    return prepend(chunk, chunks_of(n, remaining))


if __name__ == '__main__':
    # Test cases
    result = chunks_of(2, Node.from_list([1, 2, 3, 4, 5]))
    chunks = to_python_list(result)
    assert to_python_list(chunks[0]) == [1, 2]
    assert to_python_list(chunks[1]) == [3, 4]
    assert to_python_list(chunks[2]) == [5]

    result = chunks_of(3, Node.from_list([1, 2]))
    chunks = to_python_list(result)
    assert to_python_list(chunks[0]) == [1, 2]

    assert to_python_list(chunks_of(2, EMPTY_LIST)) == []

    print("All tests passed!")
